pkg/csi/service/wcp/controller.go

```diff
@@ -1938,3 +1949,57 @@ func (c *controller) ControllerModifyVolume(ctx context.Context, req *csi.Contro
 	*csi.ControllerModifyVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "")
 }
+
+func (c *controller) UpdateCNSVolumeInfo(ctx context.Context, cnsSnapshotInfo *cnsvolume.CnsSnapshotInfo) error {
+	log := logger.GetLogger(ctx)
+	var patch map[string]interface{}
+	cnsVolumeInfo, err := volumeInfoService.GetVolumeInfoForVolumeID(ctx, cnsSnapshotInfo.SourceVolumeID)
+	if err != nil {
+		return logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to retrieve cnsVolumeInfo for volume: %s Error: %+v", cnsVolumeInfo.Spec.VolumeID, err)
+	}
+	if cnsSnapshotInfo.AggregatedSnapshotCapacityInMb == -1 {
+		log.Infof("Couldn't retrieve aggregated snapshot capacity for volume %d and snapshot %d",
+			cnsVolumeInfo.Spec.VolumeID, cnsSnapshotInfo.SnapshotID)
+		patchAnnotation := common.MergeMaps(cnsVolumeInfo.Annotations,
+			map[string]string{"csi.vsphere.missing-snapshot-aggregated-capacity": "true"})
+		patch = map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"annotations": patchAnnotation,
+			},
+			"spec": map[string]interface{}{
+				"validaggregatedsnapshotsize": false,
+			},
+		}
+	} else if cnsSnapshotInfo.AggregatedSnapshotCapacityInMb != 0 {
+		log.Infof("retrived aggregated snapshot capacity for volume %d and snapshot %d",
+			cnsVolumeInfo.Spec.VolumeID, cnsSnapshotInfo.SnapshotID)
+		delete(cnsVolumeInfo.Annotations, "csi.vsphere.missing-snapshot-aggregated-capacity")
+		if cnsVolumeInfo.Spec.SnapshotLatestOperationCompleteTime.Time.Before(cnsSnapshotInfo.SnapshotLatestOperationCompleteTime) {
```

Comment:
This is misplaced, IMO. The function `UpdateCNSVolumeInfo` should simply update the CNSVolumeInfo; it shouldn't decide whether it needs to be updated, that's a job for the caller of `UpdateCNSVolumeInfo`.